server/src/main/java/io/crate/execution/dml/upsert/ShardUpsertRequest.java

```diff
@@ -443,7 +427,7 @@ public Item(StreamInput in, @Nullable Streamer<?>[] insertValueStreamers) throws
                 }
             }
             if (in.readBoolean()) {
-                source = in.readBytesReference();
+                throw new IllegalArgumentException("Cannot handle upsert request from a node running version " + in.getVersion());
```

Comment:
It's tricky to test this as part of the tests here, so I plan on looking at adding a test in `crate-qa` to catch issues here.

Reply:
We have `ShardUpsertRequestTest` where we can set arbitrary versions to test this.

Reply:
The trickiness is that unless I duplicate the StreamOutput logic for ShardUpsertRequest in the test (which will be very fragile to further updates) then there's no way of creating an older wire version - the current code always sends `false` at this point, so this exception won't get thrown.

Reply:
> the current code always sends false at this point
  
  I see. 
  
  I wonder if we could imitate old requests with non-null source via smth like 
  `item.source(new BytesArray("dummy"));`

Reply:
This PR explicitly removes the source field from `Item` though...

Reply:
Ah you are right, I was looking into my local version. Sorry for the noise.